core/app/models/spree/promotion.rb

```diff
@@ -115,8 +139,15 @@ def products
       rules.where(type: "Spree::Promotion::Rules::Product").map(&:products).flatten.uniq
     end
 
-    def usage_limit_exceeded?(promotable)
-      usage_limit.present? && usage_limit > 0 && adjusted_credits_count(promotable) >= usage_limit
+    def usage_limit_exceeded?(promotable, coupon_code = nil)
```

Comment:
if we're going to be ultimately asking a coupon code for its usage limit, it seems better to call directly on the code, rather than indirectly through the promotion here. same goes for eligible and expired.